From 31467b87c3ed5d787bb761282330e8e47962935a Mon Sep 17 00:00:00 2001
From: Tom Anderson <thomasanderson@chromium.org>
Date: Mon, 26 Sep 2022 21:20:41 +0000
Subject: [PATCH] Use system moc when not using sysroot

When using the system QT headers, we should be using the system moc too.

R=thestig

Bug: 1365358
Change-Id: Id48d4da27ad1f8154ddb0f45969be7023e0999c7
Reviewed-on: https://chromium-review.googlesource.com/c/chromium/src/+/3916120
Commit-Queue: Thomas Anderson <thomasanderson@chromium.org>
Reviewed-by: Lei Zhang <thestig@chromium.org>
Auto-Submit: Thomas Anderson <thomasanderson@chromium.org>
Cr-Commit-Position: refs/heads/main@{#1051405}
---
 ui/qt/BUILD.gn       | 22 ++++++++++++++++++----
 ui/qt/moc_wrapper.py | 10 ++++++++++
 2 files changed, 28 insertions(+), 4 deletions(-)
 create mode 100755 ui/qt/moc_wrapper.py

Index: beta/ui/qt/BUILD.gn
===================================================================
--- beta.orig/ui/qt/BUILD.gn
+++ beta/ui/qt/BUILD.gn
@@ -4,6 +4,7 @@
 
 import("//build/config/chromecast_build.gni")
 import("//build/config/linux/pkg_config.gni")
+import("//build/config/sysroot.gni")
 import("//ui/qt/qt.gni")
 
 assert(use_qt)
@@ -46,6 +47,15 @@ source_set("qt_interface") {
   sources = [ "qt_interface.cc" ]
 }
 
+if (!use_sysroot) {
+  action("generate_moc") {
+    script = "moc_wrapper.py"
+    inputs = [ "//ui/qt/qt_shim.h" ]
+    outputs = [ "$root_gen_dir/qt_shim_moc.cc" ]
+    args = rebase_path(inputs + outputs, root_build_dir)
+  }
+}
+
 shared_library("qt5_shim") {
   visibility = [
     ":qt",
@@ -72,14 +82,15 @@ shared_library("qt5_shim") {
   sources = [
     "qt_shim.cc",
     "qt_shim.h",
-
-    # This file is generated to avoid a dependency on `moc` during the build.
-    # 1. $ build/linux/debian_bullseye_amd64-sysroot/usr/lib/qt5/bin/moc \
-    #      ui/qt/qt_shim.h > ui/qt/qt_shim_moc.cc
-    # 2. $ git cl format
-    # 3. Manually add copyright header.
-    "qt_shim_moc.cc",
   ]
+  if (use_sysroot) {
+    # This file is generated with gen_qt_shim_moc.sh on an amd64 system to
+    # avoid a build-time dependency on `moc` when using the sysroot.
+    sources += [ "qt_shim_moc.cc" ]
+  } else {
+    sources += get_target_outputs(":generate_moc")
+    deps += [ ":generate_moc" ]
+  }
 }
 
 component("qt") {
Index: beta/ui/qt/moc_wrapper.py
===================================================================
--- /dev/null
+++ beta/ui/qt/moc_wrapper.py
@@ -0,0 +1,10 @@
+#!/usr/bin/env python3
+# Copyright 2022 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import subprocess
+import sys
+
+
+subprocess.check_call(["moc", sys.argv[1], "-o", sys.argv[2]])
